File: java/KeyGenerationTasks.java
import java.security.NoSuchAlgorithmException;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;

abstract class KeyGenerationTask implements Runnable {
    protected KeyGenerator keygenerator;

    @Override
    public void run(){
        for(int i=0; i<Pipeline.totalKeys; i++){
            SecretKey key = this.generateKey();
            Pipeline.keyQueue.add(key);
        }
    }
    
    abstract protected SecretKey generateKey();
}

// Concrete Key Generation Tasks

// DES

class DESKeyGeneration extends KeyGenerationTask{
    public DESKeyGeneration() throws NoSuchAlgorithmException {
        this.keygenerator = KeyGenerator.getInstance("DES");
    }
    @Override
    protected SecretKey generateKey() {
        return keygenerator.generateKey();
    }
}

// DES3

class DES3KeyGeneration extends KeyGenerationTask{
    public DES3KeyGeneration() throws NoSuchAlgorithmException {
        this.keygenerator = KeyGenerator.getInstance("DESede");
    }
    @Override
    protected SecretKey generateKey() {
        return keygenerator.generateKey();
    }
}

// AES

class AESKeyGeneration extends KeyGenerationTask{
    public AESKeyGeneration() throws NoSuchAlgorithmException {
        this.keygenerator = KeyGenerator.getInstance("AES");
    }
    @Override
    protected SecretKey generateKey() {
        return keygenerator.generateKey();
    }
}

File: python/KeyGenerationTasks.py

import sys
import crypto
sys.modules['Crypto'] = crypto
from crypto.Random import get_random_bytes
from crypto.Cipher import DES3
import Pipeline as P
from abc import ABC, abstractmethod


class KeyGenerationTask(ABC):
    def run(self):
        for i in range(P.Pipeline.total_keys):
            key = self._generate_key()
            P.Pipeline.key_queue.put(key)

    @abstractmethod
    def _generate_key(self):
        pass


# Concrete Key Generation Tasks

# DES

class DESKeyGeneration(KeyGenerationTask):
    def _generate_key(self):
        return get_random_bytes(8)

# DES3

class DES3KeyGeneration(KeyGenerationTask):
    def _generate_key(self):
        return DES3.adjust_key_parity(get_random_bytes(24))

# AES

class AESKeyGeneration(KeyGenerationTask):
    def _generate_key(self):
        return get_random_bytes(16)
